roachtest: tpcc/multiregion/survive=region/chaos=true failed

roachtest.tpcc/multiregion/survive=region/chaos=true [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3154228&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3154228&tab=artifacts#/tpcc/multiregion/survive=region/chaos=true) on master @ [ba689f91a5bbd7737a8a229522048b2ba91b2ec0](https://github.com/cockroachdb/cockroach/commits/ba689f91a5bbd7737a8a229522048b2ba91b2ec0):


```
		  |   | main.(*tpccChaosEventProcessor).checkUptime
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:48
		  |   | main.(*tpccChaosEventProcessor).listen.func1
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:265
		  |   | runtime.goexit
		  |   | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		  | Wraps: (4) expected 0 errors, found from 3763874.000000, to 3763875.000000
		  | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
		Wraps: (3) secondary error attachment
		  | error at from 2021-07-07T12:30:29Z, to 2021-07-07T12:35:19Z on metric workload_tpcc_payment_error_total{instance="34.139.41.226:2120"}: expected 0 errors, found from 3752976.000000, to 3752977.000000
		  | (1) attached stack trace
		  |   -- stack trace:
		  |   | main.(*tpccChaosEventProcessor).checkMetrics
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:190
		  |   | [...repeated from below...]
		  | Wraps: (2) error at from 2021-07-07T12:30:29Z, to 2021-07-07T12:35:19Z on metric workload_tpcc_payment_error_total{instance="34.139.41.226:2120"}
		  | Wraps: (3) attached stack trace
		  |   -- stack trace:
		  |   | main.(*tpccChaosEventProcessor).checkUptime.func2
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:65
		  |   | main.(*tpccChaosEventProcessor).checkMetrics
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:189
		  |   | main.(*tpccChaosEventProcessor).checkUptime
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:48
		  |   | main.(*tpccChaosEventProcessor).listen.func1
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:265
		  |   | runtime.goexit
		  |   | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		  | Wraps: (4) expected 0 errors, found from 3752976.000000, to 3752977.000000
		  | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
		Wraps: (4) attached stack trace
		  -- stack trace:
		  | main.(*tpccChaosEventProcessor).checkMetrics
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:190
		  | [...repeated from below...]
		Wraps: (5) error at from 2021-07-07T12:10:25Z, to 2021-07-07T12:15:15Z on metric workload_tpcc_newOrder_error_total{instance="34.139.41.226:2110"}
		Wraps: (6) attached stack trace
		  -- stack trace:
		  | main.(*tpccChaosEventProcessor).checkUptime.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:65
		  | main.(*tpccChaosEventProcessor).checkMetrics
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:189
		  | main.(*tpccChaosEventProcessor).checkUptime
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:48
		  | main.(*tpccChaosEventProcessor).listen.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/drt.go:265
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (7) expected 0 errors, found from 1889934.000000, to 1889935.000000
		Error types: (1) *secondary.withSecondaryError (2) *secondary.withSecondaryError (3) *secondary.withSecondaryError (4) *withstack.withStack (5) *errutil.withPrefix (6) *withstack.withStack (7) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh tpcc/multiregion/survive=region/chaos=true
```
</p>
</p>
</details>
/cc @cockroachdb/multiregion
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tpcc/multiregion/survive=region/chaos=true.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.tpcc/multiregion/survive=region/chaos=true [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3163908&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3163908&tab=artifacts#/tpcc/multiregion/survive=region/chaos=true) on master @ [c543d448266e2e1e3adefe3a55f920beacd848b3](https://github.com/cockroachdb/cockroach/commits/c543d448266e2e1e3adefe3a55f920beacd848b3):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/tpcc/multiregion/survive=region/chaos=true/run_1
	tpcc.go:276,tpcc.go:592,z_test_runner.go:765: error at from 2021-07-09T12:49:16Z, to 2021-07-09T12:54:06Z on metric workload_tpcc_newOrder_error_total{instance="34.75.74.69:2121"}: expected 0 errors, found from 205901.000000, to 205902.000000
		(1) secondary error attachment
		  | error at from 2021-07-09T12:59:19Z, to 2021-07-09T13:04:09Z on metric workload_tpcc_payment_error_total{instance="34.75.74.69:2122"}: expected 0 errors, found from 153805.000000, to 153806.000000
		  | (1) attached stack trace
		  |   -- stack trace:
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:190
		  |   | [...repeated from below...]
		  | Wraps: (2) error at from 2021-07-09T12:59:19Z, to 2021-07-09T13:04:09Z on metric workload_tpcc_payment_error_total{instance="34.75.74.69:2122"}
		  | Wraps: (3) attached stack trace
		  |   -- stack trace:
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime.func2
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:65
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:189
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:48
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).listen.func1
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:265
		  |   | runtime.goexit
		  |   | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		  | Wraps: (4) expected 0 errors, found from 153805.000000, to 153806.000000
		  | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:190
		  | [...repeated from below...]
		Wraps: (3) error at from 2021-07-09T12:49:16Z, to 2021-07-09T12:54:06Z on metric workload_tpcc_newOrder_error_total{instance="34.75.74.69:2121"}
		Wraps: (4) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:65
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:189
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:48
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).listen.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:265
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (5) expected 0 errors, found from 205901.000000, to 205902.000000
		Error types: (1) *secondary.withSecondaryError (2) *withstack.withStack (3) *errutil.withPrefix (4) *withstack.withStack (5) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh tpcc/multiregion/survive=region/chaos=true
```
</p>
</p>
</details>
/cc @cockroachdb/multiregion
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tpcc/multiregion/survive=region/chaos=true.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/cmd/roachtest/tests/drt.go
+++ b/pkg/cmd/roachtest/tests/drt.go
@@ -107,6 +107,23 @@ func (ep *tpccChaosEventProcessor) checkMetrics(
 	successCheckFn func(from, to model.SampleValue) error,
 	errorCheckFn func(from, to model.SampleValue) error,
 ) error {
+	// Add an extra interval to fromTime to account for the first data point
+	// which may include a node not being fully shutdown or restarted.
+	fromTime = fromTime.Add(prometheus.DefaultScrapeInterval)
+	// Similarly, scale back the toTime to account for the data point
+	// potentially already having data of a node which may have already
+	// started restarting or shutting down.
+	toTime = toTime.Add(-prometheus.DefaultScrapeInterval)
+	if !toTime.After(fromTime) {
+		l.PrintfCtx(
+			ctx,
+			"to %s < from %s, skipping",
+			toTime.Format(time.RFC3339),
+			fromTime.Format(time.RFC3339),
+		)
+		return nil
+	}
+
 	for _, check := range []struct {
 		metricType string
 		checkFn    func(from, to model.SampleValue) error
@@ -120,19 +137,6 @@ func (ep *tpccChaosEventProcessor) checkMetrics(
 			checkFn:    errorCheckFn,
 		},
 	} {
-		// Add an extra interval to fromTime to account for the first data point
-		// which may include a node not being fully shutdown or restarted.
-		fromTime := fromTime.Add(prometheus.DefaultScrapeInterval)
-		if !toTime.After(fromTime) {
-			l.PrintfCtx(
-				ctx,
-				"to %s < from %s, skipping",
-				toTime.Format(time.RFC3339),
-				fromTime.Format(time.RFC3339),
-			)
-			continue
-		}
-
 		// Verify all nodes had successes and minimal errors.
 		q := fmt.Sprintf(
 			`workload_tpcc_%s_%s_total{instance="%s:%d"}`,

--- a/pkg/cmd/roachtest/tests/drt_test.go
+++ b/pkg/cmd/roachtest/tests/drt_test.go
@@ -40,13 +40,16 @@ func TestTPCCChaosEventProcessor(t *testing.T) {
 	firstShutdownComplete := startTime.Add(95 * time.Second)
 	scrapeAfterFirstShutdownComplete := firstShutdownComplete.Add(prometheus.DefaultScrapeInterval)
 	firstPreStartup := startTime.Add(180 * time.Second)
+	scrapeBeforeFirstPreStartup := firstPreStartup.Add(-prometheus.DefaultScrapeInterval)
 	firstStartupComplete := startTime.Add(185 * time.Second)
 	scrapeAfterFirstStartupComplete := firstStartupComplete.Add(prometheus.DefaultScrapeInterval)
 
 	secondPreShutdown := startTime.Add(390 * time.Second)
+	scrapeBeforeSecondPreShutdown := secondPreShutdown.Add(-prometheus.DefaultScrapeInterval)
 	secondShutdownComplete := startTime.Add(395 * time.Second)
 	scrapeAfterSecondShutdownComplete := secondShutdownComplete.Add(prometheus.DefaultScrapeInterval)
 	secondPreStartup := startTime.Add(480 * time.Second)
+	scrapeBeforeSecondPreStartup := secondPreStartup.Add(-prometheus.DefaultScrapeInterval)
 	secondStartupComplete := startTime.Add(490 * time.Second)
 
 	metricA := "newOtan"
@@ -88,66 +91,66 @@ func TestTPCCChaosEventProcessor(t *testing.T) {
 			mockPromQueries: []expectPromQuery{
 				// Restart region1.
 				{q: makeMetric(metricA, "success", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion1), t: firstPreStartup, retVal: 100},
+				{q: makeMetric(metricA, "success", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 100},
 				{q: makeMetric(metricA, "error", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 10},
-				{q: makeMetric(metricA, "error", portRegion1), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "error", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 
 				{q: makeMetric(metricA, "success", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion2), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "success", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 				{q: makeMetric(metricA, "error", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 0},
-				{q: makeMetric(metricA, "error", portRegion2), t: firstPreStartup, retVal: 0},
+				{q: makeMetric(metricA, "error", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 0},
 
 				{q: makeMetric(metricB, "success", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion1), t: firstPreStartup, retVal: 100},
+				{q: makeMetric(metricB, "success", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 100},
 				{q: makeMetric(metricB, "error", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 10},
-				{q: makeMetric(metricB, "error", portRegion1), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "error", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 
 				{q: makeMetric(metricB, "success", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion2), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 0},
-				{q: makeMetric(metricB, "error", portRegion2), t: firstPreStartup, retVal: 0},
+				{q: makeMetric(metricB, "error", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 0},
 
 				// Shutdown region2.
 				{q: makeMetric(metricA, "success", portRegion1), t: scrapeAfterFirstStartupComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion1), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricA, "success", portRegion1), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 				{q: makeMetric(metricA, "error", portRegion1), t: scrapeAfterFirstStartupComplete, retVal: 1000},
-				{q: makeMetric(metricA, "error", portRegion1), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricA, "error", portRegion1), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 
 				{q: makeMetric(metricA, "success", portRegion2), t: scrapeAfterFirstStartupComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion2), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricA, "success", portRegion2), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 				{q: makeMetric(metricA, "error", portRegion2), t: scrapeAfterFirstStartupComplete, retVal: 1000},
-				{q: makeMetric(metricA, "error", portRegion2), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricA, "error", portRegion2), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 
 				{q: makeMetric(metricB, "success", portRegion1), t: scrapeAfterFirstStartupComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion1), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion1), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion1), t: scrapeAfterFirstStartupComplete, retVal: 1000},
-				{q: makeMetric(metricB, "error", portRegion1), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricB, "error", portRegion1), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 
 				{q: makeMetric(metricB, "success", portRegion2), t: scrapeAfterFirstStartupComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion2), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion2), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion2), t: scrapeAfterFirstStartupComplete, retVal: 1000},
-				{q: makeMetric(metricB, "error", portRegion2), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricB, "error", portRegion2), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 
 				// Restart region2.
 				{q: makeMetric(metricA, "success", portRegion1), t: scrapeAfterSecondShutdownComplete, retVal: 1000},
-				{q: makeMetric(metricA, "success", portRegion1), t: secondPreStartup, retVal: 10000},
+				{q: makeMetric(metricA, "success", portRegion1), t: scrapeBeforeSecondPreStartup, retVal: 10000},
 				{q: makeMetric(metricA, "error", portRegion1), t: scrapeAfterSecondShutdownComplete, retVal: 1000},
-				{q: makeMetric(metricA, "error", portRegion1), t: secondPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "error", portRegion1), t: scrapeBeforeSecondPreStartup, retVal: 1000},
 
 				{q: makeMetric(metricA, "success", portRegion2), t: scrapeAfterSecondShutdownComplete, retVal: 1000},
-				{q: makeMetric(metricA, "success", portRegion2), t: secondPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "success", portRegion2), t: scrapeBeforeSecondPreStartup, retVal: 1000},
 				{q: makeMetric(metricA, "error", portRegion2), t: scrapeAfterSecondShutdownComplete, retVal: 1000},
-				{q: makeMetric(metricA, "error", portRegion2), t: secondPreStartup, retVal: 10000},
+				{q: makeMetric(metricA, "error", portRegion2), t: scrapeBeforeSecondPreStartup, retVal: 10000},
 
 				{q: makeMetric(metricB, "success", portRegion1), t: scrapeAfterSecondShutdownComplete, retVal: 1000},
-				{q: makeMetric(metricB, "success", portRegion1), t: secondPreStartup, retVal: 10000},
+				{q: makeMetric(metricB, "success", portRegion1), t: scrapeBeforeSecondPreStartup, retVal: 10000},
 				{q: makeMetric(metricB, "error", portRegion1), t: scrapeAfterSecondShutdownComplete, retVal: 1000},
-				{q: makeMetric(metricB, "error", portRegion1), t: secondPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "error", portRegion1), t: scrapeBeforeSecondPreStartup, retVal: 1000},
 
 				{q: makeMetric(metricB, "success", portRegion2), t: scrapeAfterSecondShutdownComplete, retVal: 1000},
-				{q: makeMetric(metricB, "success", portRegion2), t: secondPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion2), t: scrapeBeforeSecondPreStartup, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion2), t: scrapeAfterSecondShutdownComplete, retVal: 1000},
-				{q: makeMetric(metricB, "error", portRegion2), t: secondPreStartup, retVal: 10000},
+				{q: makeMetric(metricB, "error", portRegion2), t: scrapeBeforeSecondPreStartup, retVal: 10000},
 			},
 			workloadInstances: []workloadInstance{
 				{
@@ -177,24 +180,24 @@ func TestTPCCChaosEventProcessor(t *testing.T) {
 			mockPromQueries: []expectPromQuery{
 				// Restart region1.
 				{q: makeMetric(metricA, "success", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion1), t: firstPreStartup, retVal: 100},
+				{q: makeMetric(metricA, "success", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 100},
 				{q: makeMetric(metricA, "error", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 10},
-				{q: makeMetric(metricA, "error", portRegion1), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "error", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 
 				{q: makeMetric(metricA, "success", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion2), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "success", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 				{q: makeMetric(metricA, "error", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 0},
-				{q: makeMetric(metricA, "error", portRegion2), t: firstPreStartup, retVal: 0},
+				{q: makeMetric(metricA, "error", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 0},
 
 				{q: makeMetric(metricB, "success", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion1), t: firstPreStartup, retVal: 100},
+				{q: makeMetric(metricB, "success", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 100},
 				{q: makeMetric(metricB, "error", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 10},
-				{q: makeMetric(metricB, "error", portRegion1), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "error", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 
 				{q: makeMetric(metricB, "success", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion2), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 0},
-				{q: makeMetric(metricB, "error", portRegion2), t: firstPreStartup, retVal: 1000}, // should have had errors during shutdown.
+				{q: makeMetric(metricB, "error", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000}, // should have had errors during shutdown.
 			},
 			workloadInstances: []workloadInstance{
 				{
@@ -208,7 +211,7 @@ func TestTPCCChaosEventProcessor(t *testing.T) {
 			},
 			expectedErrors: []string{
 				fmt.Sprintf(
-					`error at from 2020-12-25T00:01:45Z, to 2020-12-25T00:03:00Z on metric %s: expected 0 errors, found from 0.000000, to 1000.000000`,
+					`error at from 2020-12-25T00:01:45Z, to 2020-12-25T00:02:50Z on metric %s: expected 0 errors, found from 0.000000, to 1000.000000`,
 					makeMetric(metricB, "error", portRegion2),
 				),
 			},
@@ -230,22 +233,22 @@ func TestTPCCChaosEventProcessor(t *testing.T) {
 			mockPromQueries: []expectPromQuery{
 				// Restart region1.
 				{q: makeMetric(metricA, "success", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion1), t: firstPreStartup, retVal: 1000}, // should have had no successes whilst shutdown.
+				{q: makeMetric(metricA, "success", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 1000}, // should have had no successes whilst shutdown.
 
 				{q: makeMetric(metricA, "success", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion2), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "success", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 				{q: makeMetric(metricA, "error", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 0},
-				{q: makeMetric(metricA, "error", portRegion2), t: firstPreStartup, retVal: 0},
+				{q: makeMetric(metricA, "error", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 0},
 
 				{q: makeMetric(metricB, "success", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion1), t: firstPreStartup, retVal: 100},
+				{q: makeMetric(metricB, "success", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 100},
 				{q: makeMetric(metricB, "error", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 10},
-				{q: makeMetric(metricB, "error", portRegion1), t: firstPreStartup, retVal: 10}, // should have had errors whilst shutdown.
+				{q: makeMetric(metricB, "error", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 10}, // should have had errors whilst shutdown.
 
 				{q: makeMetric(metricB, "success", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion2), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 0},
-				{q: makeMetric(metricB, "error", portRegion2), t: firstPreStartup, retVal: 0},
+				{q: makeMetric(metricB, "error", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 0},
 			},
 			workloadInstances: []workloadInstance{
 				{
@@ -259,11 +262,11 @@ func TestTPCCChaosEventProcessor(t *testing.T) {
 			},
 			expectedErrors: []string{
 				fmt.Sprintf(
-					`error at from 2020-12-25T00:01:45Z, to 2020-12-25T00:03:00Z on metric %s: expected successes to not increase, found from 100.000000, to 1000.000000`,
+					`error at from 2020-12-25T00:01:45Z, to 2020-12-25T00:02:50Z on metric %s: expected successes to not increase, found from 100.000000, to 1000.000000`,
 					makeMetric(metricA, "success", portRegion1),
 				),
 				fmt.Sprintf(
-					`error at from 2020-12-25T00:01:45Z, to 2020-12-25T00:03:00Z on metric %s: expected errors, found from 10.000000, to 10.000000`,
+					`error at from 2020-12-25T00:01:45Z, to 2020-12-25T00:02:50Z on metric %s: expected errors, found from 10.000000, to 10.000000`,
 					makeMetric(metricB, "error", portRegion1),
 				),
 			},
@@ -287,43 +290,43 @@ func TestTPCCChaosEventProcessor(t *testing.T) {
 			mockPromQueries: []expectPromQuery{
 				// Restart region1.
 				{q: makeMetric(metricA, "success", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion1), t: firstPreStartup, retVal: 100},
+				{q: makeMetric(metricA, "success", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 100},
 				{q: makeMetric(metricA, "error", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 10},
-				{q: makeMetric(metricA, "error", portRegion1), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "error", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 
 				{q: makeMetric(metricA, "success", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion2), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricA, "success", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 				{q: makeMetric(metricA, "error", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 0},
-				{q: makeMetric(metricA, "error", portRegion2), t: firstPreStartup, retVal: 0},
+				{q: makeMetric(metricA, "error", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 0},
 
 				{q: makeMetric(metricB, "success", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion1), t: firstPreStartup, retVal: 100},
+				{q: makeMetric(metricB, "success", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 100},
 				{q: makeMetric(metricB, "error", portRegion1), t: scrapeAfterFirstShutdownComplete, retVal: 10},
-				{q: makeMetric(metricB, "error", portRegion1), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "error", portRegion1), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 
 				{q: makeMetric(metricB, "success", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion2), t: firstPreStartup, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion2), t: scrapeAfterFirstShutdownComplete, retVal: 0},
-				{q: makeMetric(metricB, "error", portRegion2), t: firstPreStartup, retVal: 0},
+				{q: makeMetric(metricB, "error", portRegion2), t: scrapeBeforeFirstPreStartup, retVal: 0},
 
 				// Shutdown region2.
 				{q: makeMetric(metricA, "success", portRegion1), t: scrapeAfterFirstStartupComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion1), t: secondPreShutdown, retVal: 100}, // should have had successes as it was not the node shutdown.
+				{q: makeMetric(metricA, "success", portRegion1), t: scrapeBeforeSecondPreShutdown, retVal: 100}, // should have had successes as it was not the node shutdown.
 
 				{q: makeMetric(metricA, "success", portRegion2), t: scrapeAfterFirstStartupComplete, retVal: 100},
-				{q: makeMetric(metricA, "success", portRegion2), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricA, "success", portRegion2), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 				{q: makeMetric(metricA, "error", portRegion2), t: scrapeAfterFirstStartupComplete, retVal: 1000},
-				{q: makeMetric(metricA, "error", portRegion2), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricA, "error", portRegion2), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 
 				{q: makeMetric(metricB, "success", portRegion1), t: scrapeAfterFirstStartupComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion1), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion1), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion1), t: scrapeAfterFirstStartupComplete, retVal: 1000},
-				{q: makeMetric(metricB, "error", portRegion1), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricB, "error", portRegion1), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 
 				{q: makeMetric(metricB, "success", portRegion2), t: scrapeAfterFirstStartupComplete, retVal: 100},
-				{q: makeMetric(metricB, "success", portRegion2), t: secondPreShutdown, retVal: 1000},
+				{q: makeMetric(metricB, "success", portRegion2), t: scrapeBeforeSecondPreShutdown, retVal: 1000},
 				{q: makeMetric(metricB, "error", portRegion2), t: scrapeAfterFirstStartupComplete, retVal: 1000},
-				{q: makeMetric(metricB, "error", portRegion2), t: secondPreShutdown, retVal: 10000}, // unexpected errors during restart.
+				{q: makeMetric(metricB, "error", portRegion2), t: scrapeBeforeSecondPreShutdown, retVal: 10000}, // unexpected errors during restart.
 			},
 			workloadInstances: []workloadInstance{
 				{
@@ -337,11 +340,11 @@ func TestTPCCChaosEventProcessor(t *testing.T) {
 			},
 			expectedErrors: []string{
 				fmt.Sprintf(
-					`error at from 2020-12-25T00:03:15Z, to 2020-12-25T00:06:30Z on metric %s: expected successes to be increasing, found from 100.000000, to 100.000000`,
+					`error at from 2020-12-25T00:03:15Z, to 2020-12-25T00:06:20Z on metric %s: expected successes to be increasing, found from 100.000000, to 100.000000`,
 					makeMetric(metricA, "success", portRegion1),
 				),
 				fmt.Sprintf(
-					`error at from 2020-12-25T00:03:15Z, to 2020-12-25T00:06:30Z on metric %s: expected 0 errors, found from 1000.000000, to 10000.000000`,
+					`error at from 2020-12-25T00:03:15Z, to 2020-12-25T00:06:20Z on metric %s: expected 0 errors, found from 1000.000000, to 10000.000000`,
 					makeMetric(metricB, "error", portRegion2),
 				),
 			},

--- a/pkg/cmd/roachtest/tests/tpcc.go
+++ b/pkg/cmd/roachtest/tests/tpcc.go
@@ -203,9 +203,14 @@ func runTPCC(ctx context.Context, t test.Test, c cluster.Cluster, opts tpccOptio
 		if err != nil {
 			t.Fatal(err)
 		}
+
 		defer func() {
+			// Use a context that will not time out to avoid the issue where
+			// ctx gets canceled if t.Fatal gets called.
+			snapshotCtx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+			defer cancel()
 			if err := p.Snapshot(
-				ctx,
+				snapshotCtx,
 				c,
 				t.L(),
 				filepath.Join(t.ArtifactsDir(), "prometheus-snapshot.tar.gz"),

--- a/pkg/cmd/roachtest/tests/tpcc.go
+++ b/pkg/cmd/roachtest/tests/tpcc.go
@@ -590,7 +590,7 @@ func registerTPCC(r registry.Registry) {
 					iter := 0
 					chaosEventCh := make(chan ChaosEvent)
 					runTPCC(ctx, t, c, tpccOptions{
-						Warehouses:     len(regions) * 5,
+						Warehouses:     len(regions) * 20,
 						Duration:       duration,
 						ExtraSetupArgs: partitionArgs,
 						ExtraRunArgs:   `--method=simple --wait=false --tolerate-errors ` + partitionArgs,

